### PR TITLE
Fix `MarkdownHeaderTextSplitter` not recognizing tilde-fenced code blocks

### PR DESCRIPTION
**Pull request by unifyh**
- **Description:** Previously `MarkdownHeaderTextSplitter` did not consider tilde-fenced code blocks (https://spec.commonmark.org/0.30/#fenced-code-blocks). This PR fixes that.
   ````md
   # Bug caused by previous implementation:
   ~~~py
   foo()
   # This is a comment that would be considered header
   bar()
   ~~~
   ````
 - **Tag maintainer:** @baskaryan

### CODE REVIEW

File: libs/langchain/langchain/text_splitter.py [hunk 1]
```diff
@@ -391,16 +391,23 @@ def split_text(self, text: str) -> List[Document]:
...
 
-            if stripped_line.startswith("```"):
-                # code block in one row
-                if stripped_line.count("```") >= 2:
+            if not in_code_block:
+                # Exclude inline code spans
+                if stripped_line.startswith("```") and stripped_line.count("```") == 1:
+                    in_code_block = True
+                    opening_fence = "```"
+                elif stripped_line.startswith("~~~"):
```
Comment (baskaryan) [inline libs/langchain/langchain/text_splitter.py:404]: do we need to do the same count check here , in case start and end of block are in a single line

Comment (unifyh) [reply libs/langchain/langchain/text_splitter.py:404]: No, see https://spec.commonmark.org/0.30/#example-146
tl;dr: if 2 `~~~`'s appear in one line, the latter is considered part of info string, instead of end of block

**Comment by unifyh** [comment]
Hi @baskaryan, is it good to go now?

**Review by hwchase17 — APPROVED** [review]

**Comment by hwchase17** [comment]
thanks!
